### PR TITLE
Add openstack as a testing platform

### PR DESCRIPTION
**Pull request by TheRealFalcon**
## Proposed Commit Message

```
Add openstack as a testing platform
```

## Test Steps
set `OPENSTACK_NETWORK` appropriately in user_settings.py. Then
`CLOUD_INIT_PLATFORM=openstack CLOUD_INIT_OS_IMAGE=3709d596-ce0d-45ac-976a-9d2f3ce46474 pytest tests/integration_tests`

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

File: tests/integration_tests/clouds.py [hunk 1]
```diff
@@ -309,3 +318,27 @@ def _get_or_set_profile_list(self, release):
...
+        )
+
+    def _get_initial_image(self):
+        image = ImageSpecification.from_os_image()
+        try:
+            UUID(image.image_id)
+        except ValueError as e:
+            raise Exception(
+                'When using Openstack, `OS_IMAGE` parameter MUST take the '
+                'form of a 36-character UUID. Passing in a release name is '
```
Comment (OddBloke) [inline tests/integration_tests/clouds.py:340]: OS_IMAGE can still take its `::` separated form (and arguably _should_, given we can't possibly infer OS/release at the moment), it's just the first component of that form which is constrained in this way. Something along the lines of "OS_IMAGE MUST be specified with a 36-character UUID image ID", perhaps?

File: tests/integration_tests/integration_settings.py [hunk 2]
```diff
@@ -87,6 +88,13 @@
 # for the keypair.  (Defaults to pycloudlib's default behaviour.)
 KEYPAIR_NAME = None
 
+##################################################################
+# OPENSTACK SETTINGS
+##################################################################
+# Network to use for Openstack. Should be one of the names/ids found
+# in `openstack network list`
+OPENSTACK_NETWORK = None
```
Comment (OddBloke) [inline tests/integration_tests/integration_settings.py:96]: If left at its default value, this produces:

```pytb
Traceback (most recent call last):
  File "/home/daniel/dev/cloud-init/tests/integration_tests/conftest.py", line 224, in client
    with _client(request, fixture_utils, session_cloud) as client:
  File "/usr/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/home/daniel/dev/cloud-init/tests/integration_tests/conftest.py", line 212, in _client
    with session_cloud.launch(
  File "/home/daniel/dev/cloud-init/tests/integration_tests/clouds.py", line 170, in launch
    pycloudlib_instance = self._perform_launch(kwargs)
  File "/home/daniel/dev/cloud-init/tests/integration_tests/clouds.py", line 144, in _perform_launch
    pycloudlib_instance = self.cloud_instance.launch(**launch_kwargs)
  File "/home/daniel/dev/pycloudlib/pycloudlib/openstack/cloud.py", line 98, in launch
    net = self.conn.network.find_network(self.network)
  File "/home/daniel/.virtualenvs/cloud-init/lib/python3.8/site-packages/openstack/network/v2/_proxy.py", line 1376, in find_network
    return self._find(_network.Network, name_or_id,
  File "/home/daniel/.virtualenvs/cloud-init/lib/python3.8/site-packages/openstack/proxy.py", line 369, in _find
    return resource_type.find(self, name_or_id,
  File "/home/daniel/.virtualenvs/cloud-init/lib/python3.8/site-packages/openstack/resource.py", line 1901, in find
    return match.fetch(session, **params)
  File "/home/daniel/.virtualenvs/cloud-init/lib/python3.8/site-packages/openstack/resource.py", line 1450, in fetch
    request = self._prepare_request(requires_id=requires_id,
  File "/home/daniel/.virtualenvs/cloud-init/lib/python3.8/site-packages/openstack/network/v2/_base.py", line 22, in _prepare_request
    req = super(NetworkResource, self)._prepare_request(
  File "/home/daniel/.virtualenvs/cloud-init/lib/python3.8/site-packages/openstack/resource.py", line 1136, in _prepare_request
    raise exceptions.InvalidRequest(
openstack.exceptions.InvalidRequest: Request requires an ID but none was found
```

That this error message is indicating that one should set `OPENSTACK_NETWORK` to something is not immediately obvious: should we introduce a guard somewhere that prompts people to configure this correctly before we reach this traceback?

**Review by OddBloke — COMMENTED** [review]
I've successfully run a test! A couple of small "UI" things I hit while getting that working, will review properly next.

File: tests/integration_tests/clouds.py (hunk 1 above)
Comment (OddBloke) [reply tests/integration_tests/clouds.py:340]: This still needs addressing, I think?

File: tests/integration_tests/integration_settings.py (hunk 2 above)
Comment (OddBloke) [reply tests/integration_tests/integration_settings.py:96]: This too, from a quick test locally.

**Comment by TheRealFalcon** [comment]
@OddBloke Sorry, not looking for another review on this one yet. I'm waiting to for the pycloudlib one to land before fixing this one up.

**Comment by OddBloke** [comment]
Drive-by thought before a meeting: do we need to expand the environment variables that tox will allow through for the OpenStack library to find credentials?

File: tests/integration_tests/clouds.py [hunk 3]
```diff
@@ -311,3 +320,32 @@ def _get_or_set_profile_list(self, release):
...
+
+    def _get_initial_image(self):
+        image = ImageSpecification.from_os_image()
+        try:
+            UUID(image.image_id)
+        except ValueError as e:
+            raise Exception(
+                'When using Openstack, `OS_IMAGE` MUST be specified with '
+                'a 36-character UUID image ID. Passing in a release name is '
+                'not valid here.\n'
```
Comment (OddBloke) [inline tests/integration_tests/clouds.py:348]: Nice clarification!

**Review by OddBloke — APPROVED** [review]
LGTM, thanks!

I'm seeing a few test failures locally, some of which are because I'm not using the latest cloud-init dev version; regardless, we can clean up any rough edges in follow-ups.

**Comment by TheRealFalcon** [comment]
> Drive-by thought before a meeting: do we need to expand the environment variables that tox will allow through for the OpenStack library to find credentials?

I don't think so. We've wildcarded the `CLOUD_INIT_*` env vars, so for an openstack test, the network would get passed as `CLOUD_INIT_OPENSTACK_NETWORK`.

**Comment by OddBloke** [comment]
> > Drive-by thought before a meeting: do we need to expand the environment variables that tox will allow through for the OpenStack library to find credentials?
> 
> I don't think so. We've wildcarded the `CLOUD_INIT_*` env vars, so for an openstack test, the network would get passed as `CLOUD_INIT_OPENSTACK_NETWORK`.

Maybe this is another oddity of my workflow, but I source my Nova credentials (populating a bunch of OS_* envvars); #830 fixes it for me locally.

**Comment by TheRealFalcon** [comment]
Oooooh, right. Yeah, I use env vars too.

**Pull request merged**